docs/source/10_deployment/12_dask.md

```diff
@@ -0,0 +1,313 @@
+
+    def create_default_data_set(self, ds_name: str) -> _DaskDataSet:
+        """Factory method for creating the default dataset for the runner.
+
+        Args:
+            ds_name: Name of the missing dataset.
+
+        Returns:
+            An instance of ``_DaskDataSet`` to be used for all
+            unregistered datasets.
+        """
+        return _DaskDataSet(ds_name)
+
+    def _run_node(
+        self,
+        node: Node,
+        catalog: DataCatalog,
+        is_async: bool = False,
+        run_id: str = None,
+        *dependencies: Node,
+    ) -> Node:
+        """Run a single `Node` with inputs from and outputs to the `catalog`.
+
+        Wraps ``run_node`` to accept the set of ``Node``s that this node
+        depends on. When ``dependencies`` are futures, Dask ensures that
+        the upstream node futures are completed before running ``node``.
+
+        Args:
+            node: The ``Node`` to run.
+            catalog: A ``DataCatalog`` containing the node's inputs and outputs.
+            is_async: If True, the node inputs and outputs are loaded and saved
+                asynchronously with threads. Defaults to False.
+            run_id: The id of the pipeline run.
+            dependencies: The upstream ``Node``s to allow Dask to handle
+                dependency tracking. Their values are not actually used.
+
+        Returns:
+            The node argument.
+        """
+        return run_node(node, catalog, is_async, run_id)
+
+    def _run(
+        self, pipeline: Pipeline, catalog: DataCatalog, run_id: str = None
+    ) -> None:
+        nodes = pipeline.nodes
+        load_counts = Counter(chain.from_iterable(n.inputs for n in nodes))
+        node_dependencies = pipeline.node_dependencies
+        node_futures = {}
+
+        client = Client.current()
+        for node in nodes:
+            dependencies = (
+                node_futures[dependency] for dependency in node_dependencies[node]
+            )
+            node_futures[node] = client.submit(
+                self._run_node, node, catalog, self._is_async, run_id, *dependencies
+            )
+
+        for i, (_, node) in enumerate(
+            as_completed(node_futures.values(), with_results=True)
+        ):
+            self._logger.info("Completed node: %s", node.name)
+            self._logger.info("Completed %d out of %d tasks", i + 1, len(nodes))
+
+            # Decrement load counts, and release any datasets we
+            # have finished with. This is particularly important
+            # for the shared, default datasets we created above.
+            for data_set in node.inputs:
+                load_counts[data_set] -= 1
+                if load_counts[data_set] < 1 and data_set not in pipeline.inputs():
+                    catalog.release(data_set)
+            for data_set in node.outputs:
+                if load_counts[data_set] < 1 and data_set not in pipeline.outputs():
+                    catalog.release(data_set)
```

Comment:
[DASK EXPERTISE REQUESTED] I've done this in the simplest way possible, in the `as_completed` loop. However, this feels a bit inefficient, as it really could've been released on the final load (rather than waiting for the node to finish running). I think this would require a distributed counter that `_DaskDataSet` instances could modify.. is this even smart?

Reply:
Upon further consideration/reading into it last night, I've decided to leave things as is, at least for now/until receiving further feedback.

1. With regards to releasing on final load, I can't see any straightforward implementation. Any logic to further segregate I/O and node execution in this runner gets more low level, which is not ideal from a tutorial perspective. Also want to avoid the case where submitting separate jobs for I/O (rather than the overarching `_run_node`, because that could mean more communication/large data transfer between workers). Overall, lot of complexity for limited-to-no benefit, IMO.
2. The suggestion to use `delayed` instead of `Future`s from the above Dask Discourse thread doesn't seem best to me, upon further consideration. That would delegate the whole DAG to Dask, and make it harder for Kedro to control anything in between. http://matthewrocklin.com/blog/work/2017/04/19/dask-glm-2 also talks a bit about the tradeoff between simplicity and flexibility between `delayed` and `Future`s, respectively.